runtime/near-vm-runner/src/cache.rs

```diff
@@ -112,286 +100,30 @@ impl fmt::Debug for MockCompiledContractCache {
-        match cache {
-            None => Ok(compile_module(wasm_code, config)),
-            Some(cache) => match cache.get(&key).map_err(|_io_err| CacheError::ReadError)? {
-                Some(record) => deserialize_wasmer(record),
-                None => compile_and_serialize_wasmer(wasm_code, config, &key, cache),
-            },
-        }
-    }
-
-    pub(crate) fn compile_module_cached_wasmer0(
-        code: &ContractCode,
-        config: &VMConfig,
-        cache: Option<&dyn CompiledContractCache>,
-    ) -> Result<Result<wasmer_runtime::Module, CompilationError>, CacheError> {
-        let key = get_contract_cache_key(code, VMKind::Wasmer0, config);
-
-        #[cfg(not(feature = "no_cache"))]
-        return WASMER_CACHE.get_or_try_put(key, |key| {
-            compile_module_cached_wasmer_impl(*key, code.code(), config, cache)
-        });
-
-        #[cfg(feature = "no_cache")]
-        return compile_module_cached_wasmer_impl(key, code.code(), config, cache);
-    }
-}
-
-#[cfg(all(feature = "wasmer2_vm", target_arch = "x86_64"))]
-pub mod wasmer2_cache {
-    use crate::wasmer2_runner::{VMArtifact, Wasmer2VM};
-    use near_primitives::contract::ContractCode;
-    use wasmer_engine::Executable;
-
-    use super::*;
-
-    pub(crate) fn compile_module_wasmer2(
-        vm: &Wasmer2VM,
-        code: &[u8],
-        config: &VMConfig,
-    ) -> Result<wasmer_engine_universal::UniversalExecutable, CompilationError> {
-        let _span = tracing::debug_span!(target: "vm", "compile_module_wasmer2").entered();
-        let prepared_code =
-            prepare::prepare_contract(code, config).map_err(CompilationError::PrepareError)?;
-        vm.compile_uncached(&prepared_code)
-    }
-
-    pub(crate) fn compile_and_serialize_wasmer2(
-        wasm_code: &[u8],
-        key: &CryptoHash,
-        config: &VMConfig,
-        cache: &dyn CompiledContractCache,
-    ) -> Result<Result<VMArtifact, CompilationError>, CacheError> {
-        let _span = tracing::debug_span!(target: "vm", "compile_and_serialize_wasmer2").entered();
-        let vm = Wasmer2VM::new(config.clone());
-        let executable = match compile_module_wasmer2(&vm, wasm_code, config) {
-            Ok(module) => module,
-            Err(err) => {
-                cache_error(&err, key, cache)?;
-                return Ok(Err(err));
-            }
-        };
-        let code =
-            executable.serialize().map_err(|_e| CacheError::SerializationError { hash: key.0 })?;
-        let record = CompiledContract::Code(code);
-        cache.put(key, record).map_err(|_io_err| CacheError::WriteError)?;
-        match vm.engine.load_universal_executable(&executable) {
-            Ok(artifact) => Ok(Ok(Arc::new(artifact) as _)),
-            Err(err) => {
-                let err = CompilationError::WasmerCompileError { msg: err.to_string() };
-                cache_error(&err, key, cache)?;
```

Comment:
And here's the "cached" path counterpart of 

https://github.com/near/nearcore/pull/7851/files#r997413310

I change the behavor here to return a CacheError, rather than cache error. I think this is what we should do (ie, the current code on master is buggy). `load_universal_executable` fails only if we fail to allocate memory, and that's definitely not an error we should cache (@nagisa coudl you double-check this?)

Reply:
Yeah, I believe `load_universal_executable` can only fail for obscure resource reasons. The `ref` variant can additionally fail if the backing serialized storage is corrupted in some way, but that’s also something we shouldn’t be caching I feel.